src/service/angular-sails.js

```diff
@@ -1,74 +1,238 @@
     this.url = undefined;
-    this.interceptors = [];
-    this.responseHandler = undefined;
-
-    this.$get = ['$q', '$timeout', function ($q, $timeout) {
-        var socket = io.connect(provider.url),
-            defer = function () {
-                var deferred = $q.defer(),
-                    promise = deferred.promise;
-
-                promise.success = function (fn) {
-                    promise.then(fn);
-                    return promise;
-                };
-
-                promise.error = function (fn) {
-                    promise.then(null, fn);
-                    return promise;
-                };
-
-                return deferred;
-            },
-            resolveOrReject = this.responseHandler || function (deferred, data) {
-                // Make sure what is passed is an object that has a status that is a number and if that status is no 2xx, reject.
-                if (data && angular.isObject(data) && data.status && !isNaN(data.status) && Math.floor(data.status / 100) !== 2) {
-                    deferred.reject(data);
-                } else {
-                    deferred.resolve(data);
+    
+    // Prefix every request with this url
+    this.urlPrefix = '';
+    
+    this.config = {
+      // Transports to use when communicating with the server, in the order they will be tried
+      transports: ['websocket', 'polling'],
+      useCORSRouteToGetCookie: false
+    };
+    this.debug = false;
+
+    // like https://docs.angularjs.org/api/ng/service/$http#interceptors
+    // but with sails.io arguments
+    var interceptorFactories = this.interceptors = [
+      /*function($injectables) {
+         return {
+           request: function(config) {},
+           response: function(response) {},
+           requestError: function(config) {},
+           responseError: function(response) {}
+         };
+       }*/
+    ];
+
+    this.$get = ['$q', '$injector', '$rootScope', '$log', function($q, $injector, $rootScope, $log) {
+      var socket = (io.sails || io).connect(provider.url, provider.config);
+
+      // build interceptor chain
+      var reversedInterceptors = [];
+      angular.forEach(interceptorFactories, function(interceptorFactory) {
+        reversedInterceptors.unshift(
+          angular.isString(interceptorFactory) ?
+            $injector.get(interceptorFactory) : $injector.invoke(interceptorFactory));
+      });
+
+      // Send the request using the socket
+      function serverRequest(config) {
+        var defer = $q.defer();
+        if(provider.debug) $log.info('$sails ' + config.method.toUpperCase() + ' ' + config.url, config.data || '');
+
+        socket['legacy_' + config.method](config.url, config.data, function(result, jwr) {
+          // resolve promise if JSON web response is an object that has a statusCode 2xx
+          if(!jwr) jwr = {};
+          jwr.data = result; // backward and $http compat
+          jwr.status = jwr.statusCode; // $http compat
```

Comment:
If `jwr` is falsey on line 98,` jwr.statusCode` will be undefined here. Maybe we can have some sort of fallback convention looking to see if result is an object and if that object has `status` or `statusCode`.  You end up looking at `jwr.statusCode` on 105 in the if statement to determine if an error was returned